grpc-gcp/src/main/java/com/google/grpc/gcp/GcpManagedChannel.java

```diff
@@ -185,6 +199,47 @@ private void initMetrics() {
     gauge.createTimeSeries(labelValues, obj, func);
   }
 
+  /**
+   * ChannelStateMonitor subscribes to channel's state changes and informs {@link GcpManagedChannel}
+   * on any new state. This monitor allows to detect when a channel is not ready and temporarily
+   * route requests via another ready channel if the option is enabled.
+   */
+  private class ChannelStateMonitor implements Runnable {
+    private final int channelId;
+    private final ManagedChannel channel;
+
+    private ChannelStateMonitor(ManagedChannel channel, int channelId) {
+      this.channelId = channelId;
+      this.channel = channel;
+      run();
+    }
+
+    @Override
+    public void run() {
+      if (channel == null) {
+        return;
+      }
+      ConnectivityState newState = channel.getState(false);
+      processChannelStateChange(channelId, newState);
+      if (newState != ConnectivityState.SHUTDOWN) {
+        channel.notifyWhenStateChanged(newState, this);
+      }
+    }
+  }
+
+  private void processChannelStateChange(int channelId, ConnectivityState state) {
+    if (!fallbackEnabled) {
+      return;
+    }
+    if (state == ConnectivityState.READY || state == ConnectivityState.IDLE) {
+      // Ready
+      fallbackMap.remove(channelId);
+      return;
+    }
+    // Not ready
```

Comment:
No, nothing special is needed for the SHUTDOWN state or any other state. They will be treated as "not ready" and if a new RPC will come to a channel in that state we will try to provide a fallback channel. In case of a SHUTDOWN - currently it must happen only when client shuts down -- so most likely all the channels will be in the SHUTDOWN state at that point. We will fail to provide a fallback but it doesn't matter as the client shuts down anyway.